### DISCUSSION TITLE
VisualShaderNodeCustom Improvement ideas.

### DISCUSSION THREAD

**Discussion by donn-xx** [godotengine/godot-proposals#6069] [Engine Core]
Global Code Problems
==
In *_get_global_code()* there is no way to return a string that caters for whether the global code *already* contains a function of the name you wish to return.

Edit
--
It's more subtle than this. Custom node class_name A and another class_name B that return the same global code will cause the clashing names—and there's no access in gdscript from A<-->B.

Solution:
--
1. The shader compilation could remove duplicate functions that are identical in name and content—no matter the node that supplies them.
2. Some global object/dictionary available, at the shader-level, so we can track what functions have been used by other nodes.
3. This is doable from gdscript so please ignore: The functions could be given some unique name .. eg. func "_get_global_code(uid:String)".

Port Types
==
It would be great to have a Variant port (in and out) so that it's possible to make pass-through nodes, or nodes that have a chance to convert things and pass them out.

UI Fields in slots
==
Custom nodes need combo-boxes, text boxes, sliders and other UI goodies.

Hope that made sense.

**Comment by Chaosus** [comment]
You can define a single function for multiple custom nodes using GlobalExpression node instead. I think adding an extra parser makes it too complex and may have undesirable performance penalty.

**Reply by donn-xx** [reply]
That's a good out-of-the-box idea. I'll have a think.
